### DISCUSSION TITLE
ESM output does not support ESM externals, because “the target environment doesn't support EcmaScriptModule syntax ”

### DISCUSSION THREAD

**Discussion by dilyanpalauzov** [webpack/webpack#15291] [Q&A] [answered]
# Bug report
```javascript
//this is src/index.js
import * as Popper from '@popperjs/core';

export default Popper.createPopper(null, null);
```
```javascript
//this is webpack.config.js
import webpack from 'webpack'

export default {
    optimization: {
        minimize: false,
        concatenateModules: true
    },
    experiments: {
        outputModule: true
    },
    output: {
        library: {
            type: 'module'
        }
    },
    node: { global: false },
    externals: {
        "@popperjs/core": "@popperjs/core"
    }
}
```
My package.json has `"type": "module",`.

**What is the current behavior?**
Calling `yarn webpack` prints
```
$ /home/d/A/node_modules/.bin/webpack
assets by status 136 bytes [cached] 1 asset
orphan modules 42 bytes [orphan] 1 module
./src/index.js + 1 modules 202 bytes [not cacheable] [built] [code generated]

WARNING in configuration
The 'mode' option has not been set, webpack will fallback to 'production' for this value.
Set 'mode' option to 'development' or 'production' to enable defaults for each environment.
You can also set it to 'none' to disable any default behavior. Learn more: https://webpack.js.org/configuration/mode/

ERROR in ./src/index.js + 1 modules
The target environment doesn't support EcmaScriptModule syntax so it's not possible to use external type 'module'
while analyzing module external module "@popperjs/core" for concatenation
Error: The target environment doesn't support EcmaScriptModule syntax so it's not possible to use external type 'module'
while analyzing module external module "@popperjs/core" for concatenation
    at ExternalModule._getSourceData (/home/d/A/node_modules/webpack/lib/ExternalModule.js:590:12)
    at ExternalModule.codeGeneration (/home/d/A/node_modules/webpack/lib/ExternalModule.js:650:29)
    at ConcatenatedModule._analyseModule (/home/d/A/node_modules/webpack/lib/optimize/ConcatenatedModule.js:1659:29)
    at ConcatenatedModule.codeGeneration (/home/d/A/node_modules/webpack/lib/optimize/ConcatenatedModule.js:1102:9)
    at /home/d/A/node_modules/webpack/lib/Compilation.js:3323:22
    at /home/d/A/node_modules/webpack/lib/Cache.js:93:5
    at Hook.eval [as callAsync] (eval at create (/home/d/A/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:6:1)
    at Cache.get (/home/d/A/node_modules/webpack/lib/Cache.js:75:18)
    at ItemCacheFacade.get (/home/d/A/node_modules/webpack/lib/CacheFacade.js:111:15)
    at Compilation._codeGenerationModule (/home/d/A/node_modules/webpack/lib/Compilation.js:3316:9)

webpack 5.68.0 compiled with 1 error and 1 warning in 132 ms
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Removing `externals` from webpack.config.js and calling `yarn webpack` prints:
```
$ yarn webpack
yarn run v1.22.17
$ /home/d/A/node_modules/.bin/webpack
asset main.mjs 65.2 KiB [compared for emit] [javascript module] (name: main)
orphan modules 76.5 KiB [orphan] 58 modules
runtime modules 396 bytes 2 modules
./src/index.js + 52 modules 71.4 KiB [built] [code generated]

WARNING in configuration
The 'mode' option has not been set, webpack will fallback to 'production' for this value.
Set 'mode' option to 'development' or 'production' to enable defaults for each environment.
You can also set it to 'none' to disable any default behavior. Learn more: https://webpack.js.org/configuration/mode/

webpack 5.68.0 compiled with 1 warning in 485 ms
Done in 1.06s.
```
`dist/main.mjs` uses `import` and `export`, so webpack does think, that the target environment supports EcmaScriptModule syntax.

**What is the expected behavior?**
The expected behaviour is to support ESM-output with ESM-external dependencies.  That is, webpack shall not refuse to use external type 'module', when the output is ESM, reasoning that the target does not support ESM.

```
$ /home/d/A/node_modules/.bin/webpack --version
webpack: 5.68.0
webpack-cli: 4.9.2
webpack-dev-server not installed
$ node --version
v16.13.1
```

**Comment by alexander-akait** [comment]
Because you don't have the `target` option, webpack tries to load `browserslist` file, can you have it? If yes, please provide content

**Reply by dilyanpalauzov** [reply]
The default/implicit target is `web`.  For that target, the output can be ESM.  This can be seen after removing `externals` from the above configuration and running `yarn webpack`.  That said, the default target `web` does support ESM.

When `externals` is not removed, webpack says, that the target=`web` does not support EcmaScriptModule syntax.  This statement is not correct, as the previous paragraph explains, why the `web` target supports EcmaScriptModule syntax.

**Reply by vankop** [reply]
you should pass target with es >= 2015 https://github.com/webpack/webpack/blob/main/lib/config/target.js#L276

**Reply by dilyanpalauzov** [reply]
Why shall I pass any target?  When the webpack-output is ESM, and it is in this example ESM, then webpack shall assume that the target supports ESM syntax.  Currently when the output is ESM, webpack is unsure, whether the target supports ESM syntax. There is no place for such doubts.

**Reply by alexander-akait** [reply]
Because not every env supports ESM, maybe we revisit it late (in the next major release), but currently better output the error, web can be in ES3, ES5, and other

**Reply by dilyanpalauzov** [reply]
When webpack is told to output in ESM syntax, please tweak webpack to assume, that the target support ESM syntax. (Thinking that the target does not support ESM syntax, while emitting ESM output, makes no sense).

**Reply by dilyanpalauzov** [reply]
> Because not every env supports ESM, maybe we revisit it late (in the next major release), but currently better output the error, web can be in ES3, ES5, and other

In this particular case, the environment  supports ESM, because the output is in ESM.

**Reply by alexander-akait** [reply]
> When webpack is told to output in ESM syntax, please tweak webpack to assume, that the target support ESM syntax. (Thinking that the target does not support ESM syntax, while emitting ESM output, makes no sense).

That is not truth, you can enable module but set targets on other browsers/env, we should output something on this, otherwise you will potential break your app for your clients

**Reply by dilyanpalauzov** [reply]
> That is not truth, you can enable module but set targets on other browsers/env, 

Can you describe a use case, where one wants target:ES3 or ES5 and output: module?

How can I define a target `web with ESM`, which behaves like target `web`, has es6 modules and otherwise does not change the code anyhow (as target web does not change the code)?

**Reply by alexander-akait** [reply]
You can use ModuleFederation with other target, you can use `DllPlugin` with other type, you can apply NodeJs plugins and not only, which can't support ESM in some cases, you set different chunkLoading for entries https://github.com/webpack/webpack/blob/main/types.d.ts#L3193, which can't support ESM, do you need more?

**Reply by dilyanpalauzov** [reply]
target:ES3 is mutually exclusive to output:module.  The reason is, that output:module implies target:ES6 at minimum.

**Reply by alexander-akait** [reply]
Do you read my answer above? You can set different chunk loading per entry and not all chunk loading can be used with ESM, so yes, it is possible to set module output and generate non ESM code

**Reply by dilyanpalauzov** [reply]
I read your answer, but to be honest, I could not understand it.

**Comment by dilyanpalauzov** [comment, marked as answer]
I will rephrase the problem:

When the output is `module`, webpack wrongly assumes for target `web`, that the target does not support ESM syntax.

**Reply by vankop** [reply]
we decided to enhanced defaults in this case https://github.com/webpack/webpack/issues/15292

**Reply by Jdruwe** [reply]
I am also getting this error with following config:

```
   experiments: {
        outputModule: true,
    },
    output: {
        filename: "[name].min.js",
        module: true,
        library: {
            type: "module",
        },
        globalObject: "this",
    },
    externals: {
        vue: "Vue",
    },
```

Anything missing?
